### PR TITLE
Correctly handle cases when a replicated distribution can be tainted

### PR DESCRIPTION
**Pull request by asubramanya**
When a query contains non deterministic operators on top of a replicated
distribution of tuples, the resulting distribution produced by these operators
is no longer replicated. Examples of such operators are global Limit, Window
Aggs, non trivial Aggs. They produce a "Tainted" form of replicated
distribution. ORCA does not have a way to represent this type of distribution.
In order to support this we require a "GeneralReplicated"
distribution spec and a "TaintedReplicated" distribution spec.

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/operators/CPhysicalInnerNLJoin.cpp [hunk 1]
```diff
@@ -161,7 +162,7 @@ CPhysicalInnerNLJoin::PdsRequired
 
 	if (0 == child_index)
 	{
-		return GPOS_NEW(mp) CDistributionSpecReplicated();
+		return GPOS_NEW(mp) CDistributionSpecGeneralReplicated();
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/operators/CPhysicalInnerNLJoin.cpp:165]: How are we determining whether to return a GeneralReplicated vs a regular Replicated request? In CPhysicalInnerIndexNLJoin we still return `CDistributionSpecReplicated`.

File: src/backend/gporca/data/dxl/minidump/TaintedReplicatedAgg.mdp [hunk 2]
```diff
@@ -0,0 +1,351 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+  <dxl:Comment><![CDATA[
+  CREATE TABLE snackbox (a float4) DISTRIBUTED REPLICATED;
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply src/backend/gporca/data/dxl/minidump/TaintedReplicatedAgg.mdp:4]: Will do.

File: src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h [hunk 3]
```diff
@@ -41,6 +41,8 @@ namespace gpopt
 				EdtHashedNoOp,		// same as hashed, used to force multiple slices for parallel union all. The motions always mirror the underlying distributions.
 				EdtStrictHashed,	// same as hashed, used to force multiple slices for parallel union all. The motions mirror the distribution of the output columns.
 				EdtReplicated,		// data is replicated across all segments
+				EdtGeneralReplicated,// data is replicated across all segments or is Tainted replicated (required only)
+				EdtTaintedReplicated,// data is replicated across all segments but not really
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h:45]: Take for example, a Limit or WindowAgg on top of a replicated distribution. These operators when executed on each segment might end up producing a different set of results based on the order in which they process the tuples. So the resulting distribution is no longer replicated and is now tainted.

File: src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpecGeneralReplicated.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+#define GPOPT_CDistributionSpecGeneralReplicated_H
+
+#include "gpos/base.h"
+
+#include "gpopt/base/CDistributionSpec.h"
+
+namespace gpopt
+{
+
+	// derive-only: unsafe computation over replicated data
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpecGeneralReplicated.h:14]: General should be required only.

File: src/backend/gporca/libgpopt/src/operators/CPhysicalInnerNLJoin.cpp [hunk 5]
```diff
@@ -11,6 +11,7 @@
 
 #include "gpos/base.h"
 #include "gpopt/base/CDistributionSpecReplicated.h"
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply src/backend/gporca/libgpopt/src/operators/CPhysicalInnerNLJoin.cpp:13]: Will get rid of this.

File: src/backend/gporca/libgpopt/src/base/CDistributionSpecTaintedReplicated.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#include "gpopt/base/CDistributionSpecTaintedReplicated.h"
+
+#include "gpopt/base/CDistributionSpecNonSingleton.h"
+#include "gpopt/base/CDistributionSpecSingleton.h"
+
+namespace gpopt
+{
+
+BOOL
+CDistributionSpecTaintedReplicated::FSatisfies(const CDistributionSpec *pds) const
```
Comment (hardikar) [inline src/backend/gporca/libgpopt/src/base/CDistributionSpecTaintedReplicated.cpp:13]: This "satisfies" matrix is very similar to "CDistributionSpecReplicated::FSatisfies". I was having trouble understanding this, so I summarized it into the follow table.

|                      | Replicated       | TaintedReplicated |
|----------------------|------------------|-------------------|
| Matches              | true             | (default) false   |
| NonSingleton         | FAllowReplicated | same              |
| GeneralReplicated    | true             | same              |
| singleton & master   | false            | same              |
| singleton & segment  | true             | same              |
| ANY                  | true             | same              |
| others not Singleton | true             | (default) false   |

So it looks like the main difference between the two is that Replicated satisfies Replicated; TaintedReplicated doesn't satisfy itself or Replicated. I'm not sure how the last line in the table above is relevant, probably not, but it's odd that Replicated would satisfy Hashed distribution spec. If that's correct, should TaintedReplicated as well?

Comment (hardikar) [reply src/backend/gporca/libgpopt/src/base/CDistributionSpecTaintedReplicated.cpp:13]: In either case, please highlight the fact that *the main difference between the two is that Replicated satisfies Replicated; TaintedReplicated doesn't satisfy itself or Replicated.* in a comment.

File: src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h (hunk 3 above)
Comment (hardikar) [inline src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h:45]:
```suggestion
				EdtTaintedReplicated,   // data once-replicated, after being processed by an input-order-sensitive operator
```

File: src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h [hunk 7]
```diff
@@ -41,6 +41,8 @@ namespace gpopt
 				EdtHashedNoOp,		// same as hashed, used to force multiple slices for parallel union all. The motions always mirror the underlying distributions.
 				EdtStrictHashed,	// same as hashed, used to force multiple slices for parallel union all. The motions mirror the distribution of the output columns.
 				EdtReplicated,		// data is replicated across all segments
+				EdtGeneralReplicated,// data is replicated across all segments or is Tainted replicated (required only)
```
Comment (hardikar) [inline src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h:44]:
```suggestion
				EdtGeneralReplicated,   // data is strict or tainted replicated (required only)
```

File: src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h [hunk 8]
```diff
@@ -41,6 +41,8 @@ namespace gpopt
 				EdtHashedNoOp,		// same as hashed, used to force multiple slices for parallel union all. The motions always mirror the underlying distributions.
 				EdtStrictHashed,	// same as hashed, used to force multiple slices for parallel union all. The motions mirror the distribution of the output columns.
 				EdtReplicated,		// data is replicated across all segments
```
Comment (hardikar) [inline src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h:43]:
```suggestion
				EdtReplicated,		// data is strictly replicated across all segments
```

**Comment by d** [comment]
Thanks @zellerh for your time reviewing this!

> General comment: There are at least three places where we can't use replicated: a) global limit, b) window functions that are order-dependent and c) operators that contain volatile functions. This PR tries to solve a) and b). There is already logic to handle c). For example, see `CPhysicalComputeScalar::PdsRequired`. This uses a different method, probably a bit too restrictive, by requiring singleton execution. But, I don't think we should have two totally different methods to solve the same problem, so my suggestion would be to either solve a) and b) the existing way or to change c) to use your new method.
> 
I agree with you, I think the existing method in c) is too restrictive, and the finer-grained distribution spec should allow us to get better plans for cases handled by (c).

> My gut feeling tells me that a better solution might be to leave derived properties the way they are and to solve this problem by letting limit and window agg generate a required distribution spec that disallows replication. The same method could then be used for all three use cases.

To be clear, before this PR, ORCA basically is conflating two kinds of "replicated"

1. "Pure" replicated, where things just went through a Broadcast motion, where the relation is _identical_ across segments
1. "Tainted" replicated, where we perform a class of relational operations on top of something either purely or tainted-ly (recursively defined here) replicated. Example operations are: `LIMIT`, "Group By Aggregate", "Aggregate", "Window Functions".

One alternative we have briefly entertained was to outlaw queries that are under-specified or has undefined outcome. But I quickly realized that SQL standard defines "undefined behavior" with _more than just the query_, a query has _undefined_ behavior when its data leads to ambiguous outcomes. This makes it _impossible_ for an optimizer to actually rule out UB in queries. For example, if table `grandma` has one or zero tuples, the following query is well-defined:

   ```sql
   SELECT (SELECT a FROM grandma) FROM me;
   ```

Another alternative we entertained was as you suggested, to outlaw "tainted" replicated, but that seems to come with a pretty hefty price at optimality. Because in order to keep replicated "pure", we need to make sure the following relational operations ban replicated child:

  1. Limit (specifically global limit)

  1. SequenceProject (a.k.a. windowing functions)

  1. Agg. Yeah this one is a shocker, see example below

     ```sql
     CREATE TEMP TABLE grandma (a) AS (VALUES ((2 ^ -102)::float4), (-(2 ^ -103)::float4), ((2 ^ -126)::float4));
     
     SELECT sum(a) FROM (SELECT a FROM grandma ORDER BY random()) t;
     ```

     I'm using `random()` to simulate different scan order here. The two plausible results are:

     1. `2 ^ -103` if the largest number is first added with the smallest, losing precision, then subtracting the medium number. Essentially, the small number is "lost"

        ```
             sum      
        --------------
         9.860762e-32
        (1 row)
        ```

     1.  `2 ^ -103 + 2^ - 126` if we perform the addition of the largest number with the "medium" (in absolute value) number first, preserving precision.

        ```
              sum      
        ---------------
         9.8607613e-32
        (1 row)
        ```

     Now the above might be a contrived example, but unless we take an "allow list" approach, we'll have to give up on running aggregates over broadcast. Which sounds like easy and no big deal. But the ripple effect is that we might be forced to broadcast "the other side" of joins which used to be just partitioned.

**Comment by zellerh** [comment]
Thanks a lot for the detailed explanation, @d. It's always good to establish the facts and your explanation explains those very well.

My vote would be to outlaw "tainted" replication completely. Here are what I would suggest as the rules. I would suggest that we disallow execution of the following in a replicated part of a slice:

1. Global limit - as you mentioned.
2. SequenceProject - as you mentioned. We still can hash-distribute on a subset of the PARTITION BY columns, if any are specified.
3. Some aggregate functions. One that @asubramanya mentioned is array_agg. One might also need to outlaw aggregates on floating point expressions, as you explain. The vast majority of aggregates like sum, count, min, max, avg should be allowed.
4. Operators with volatile functions in them.

Given that each of the above can cause query results that could not be produced with serial execution, I feel that disallowing the scenarios is the right choice.

Overall, my guiding principle is this: Users of GPDB may sometimes write a query that gives non-deterministic results, and they need to be prepared to deal with them. However, that does not mean that we as system implementors can return results that would clearly be labeled as incorrect in a non-parallel system.

I hope that we can get more than two people in this community to voice an opinion and will be ok to go with what a majority decides.

**Comment by d** [comment]
On Aug 3, 2020, 4:21 PM PDT @zellerh wrote:
> Thanks a lot for the detailed explanation, @d. It's always good to establish the facts and your explanation explains those very well.
> 
> My vote would be to outlaw "tainted" replication completely. Here are what I would suggest as the rules. I would suggest that we disallow execution of the following in a replicated part of a slice:
> 
> 1. Global limit - as you mentioned.
> 2. SequenceProject - as you mentioned. We still can hash-distribute on a subset of the PARTITION BY columns, if any are specified.
> 3. Some aggregate functions. One that @asubramanya mentioned is `array_agg`. One might also need to outlaw aggregates on floating point expressions, as you explain. The vast majority of aggregates like sum, count, min, max, avg should be allowed.

My take would be that we should assume _all_ aggregates taint replication by default. But I'd like to hear what you suggest. How will ORCA recognize the "vast majority of aggregates"? Would you recommend a "disallow list" or an "allow list"? Where do user-defined aggregates fit into this characterization? As the system evolves, will we need to remember to expand the list, once we decide whether we do an "allow" list or one that disallows?

> 4. Operators with volatile functions in them.
> 
> Given that each of the above can cause query results that could not be produced with serial execution, I feel that disallowing the scenarios is the right choice.
> 
> Overall, my guiding principle is this: Users of GPDB may sometimes write a query that gives non-deterministic results, and they need to be prepared to deal with them. However, that does not mean that we as system implementors can return results that would clearly be labeled as incorrect in a non-parallel system.

Thank you for bring up the contrast to a non-parallel system. It's a place I keep visiting while thinking about this topic.

> would clearly be labeled as incorrect in a non-parallel system.

Can you illustrate this point with an example of an execution plan under this proposal that will lead to _clearly incorrect_ outcome in a parallel system?

**Comment by zellerh** [comment]
> My take would be that we should assume _all_ aggregates taint replication by default. But I'd like to hear what you suggest. How will ORCA recognize the "vast majority of aggregates"? Would you recommend a "disallow list" or an "allow list"? Where do user-defined aggregates fit into this characterization? As the system evolves, will we need to remember to expand the list, once we decide whether we do an "allow" list or one that disallows?

An "allow" list might be easier. To start with, I think it is safe to put the following on the allow list: MIN, MAX for numeric (both exact and approximate), datetime, text. SUM and AVG for exact numeric, COUNT.

> Can you illustrate this point with an example of an execution plan under this proposal that will lead to _clearly incorrect_ outcome in a non-parallel system?

Here is an example:

```
drop table if exists sales, prod;
create table sales (id int, prod_cd int, qty int);
create table prod  (prod_cd int, prod_descr text);

insert into sales select i, i%100, 3 from generate_series(1,1000000) i;
insert into prod  select i, 'prod_' || i::text from generate_series(0,99) i;

-- take an arbitrary product and sum up its sales
explain
select prod_descr, sum(qty) from sales join (select * from prod limit 1) p on sales.prod_cd = p.prod_cd group by prod_descr;
```

When I explain this on 6X, I get a broadcast join, but the limit is executed on a single segment, so we get the expected answer: A single product and its sales of 30000. If we would have computed the limit in a replicated portion of a slice, the result may show more than one product and the sales number(s) would be incorrect (!= 30000) in that case. To me, such a result - more than one row, with partial sums - would be clearly incorrect.

**Comment by zellerh** [comment]
Expected result of this query (could be any product):

```
 prod_descr |  sum
------------+-------
 prod_5     | 30000
(1 row)
```

Example of an incorrect result with replicated execution of the limit:

```
 prod_descr |  sum
------------+-------
 prod_5     | 20123
 prod_0     |  9956
(2 rows)
```

**Comment by d** [comment]
I wrote:
> Can you illustrate this point with an example of an execution plan under this proposal that will lead to _clearly incorrect_ outcome in a parallel system?

On Aug 4, 2020, 11:07 AM PDT @zellerh wrote:
> Here is an example:
> 
> ```
> drop table if exists sales, prod;
> create table sales (id int, prod_cd int, qty int);
> create table prod  (prod_cd int, prod_descr text);
> 
> insert into sales select i, i%100, 3 from generate_series(1,1000000) i;
> insert into prod  select i, 'prod_' || i::text from generate_series(0,99) i;
> 
> -- take an arbitrary product and sum up its sales
> explain
> select prod_descr, sum(qty) from sales join (select * from prod limit 1) p on sales.prod_cd = p.prod_cd group by prod_descr;
> ```
> 
> When I explain this on 6X, I get a broadcast join, but the limit is executed on a single segment, so we get the expected answer: A single product and its sales of 30000. If we would have computed the limit in a replicated portion of a slice, the result may show more than one product and the sales number(s) would be incorrect (!= 30000) in that case. To me, such a result - more than one row, with partial sums - would be clearly incorrect.

Thanks for this example. However, I do not believe this query above correctly expresses the intent of "take a single sample of a table". I'd like to emphasize that a (non-parallel) database is permitted to emit the following plan:

```
 HashAggregate
   Group Key: prod.prod_descr
   ->  Nested Loop
         Join Filter: (sales.prod_cd = prod.prod_cd)
         ->  Seq Scan on sales
         ->  Limit
               ->  Seq Scan on prod
```

In the plan above, we're not guaranteed to match the same tuple from prod with every tuple in sales.

To be clear, I want to be cognizant of the risk of disappointing a user when a query they write doesn't do what they expect. But I'll need a better example than the above query, because here the expectation is wrong. In fact, I'm inclined to assert such an intent is *impossible* to express in SQL. If it worked on Greenplum or Postgres it was likely coincidental. IMHO this is a task that's much better suited for a procedural language like PL/pgSQL or PL/SQL, and I'd accordingly recommend a rewrite of the above query into the following:

```PLpgSQL
CREATE FUNCTION procedural_query() RETURNS TABLE(prod_desc text, sum bigint)
VOLATILE LANGUAGE plpgsql AS $codeblock$
DECLARE
p prod%ROWTYPE;
BEGIN
SELECT * INTO p FROM prod LIMIT 1;
RETURN QUERY
SELECT p.prod_descr, sum(qty)
FROM sales
WHERE sales.prod_cd = p.prod_cd;
END;
$codeblock$;

SELECT * FROM procedural_query();
```

**Comment by zellerh** [comment]
Thanks, @d, I just looked at the SQL 2003 and SQL 1999 standards and based on what they say, you are right. In the description of the <from clause>, it explicitly says that it is executed in a nested way. Actually, I am surprised to see that and my interpretation probably shows my age, as the older SQL93 standard has a different way to define a <from clause> as the cartesian product of the <table reference>s in it. Probably this was changed when the `LATERAL` join was introduced. We could still go with my stricter interpretation if we wanted to. Now, following the SQL 2003 standard, the following query would actually **require** us to always use a nested loop join for a query like this:

`select * from foo, (select * from bar where my_volatile_func(bar.a) = 10)`

Do we want to guarantee this behavior?

Maybe what we should do is a bit of investigation of how Postgres and other popular DBMSs behave with these types of non-deterministic queries.

**Comment by d** [comment]
> Maybe what we should do is a bit of investigation of how Postgres and other popular DBMSs behave with these types of non-deterministic queries.

I was part of a short [conversation in upstream PostgreSQL](https://www.postgresql.org/message-id/flat/SN6PR05MB4559835263BADF54FA370163B57E0%40SN6PR05MB4559.namprd05.prod.outlook.com) a week or so ago, which is quite illuminating.

**Comment by asubramanya** [comment]
Closing this PR since it was decided to go with the approach suggested by @zellerh.

**Pull request closed without merging**